core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java

```diff
@@ -817,19 +896,94 @@ public boolean isCompatible(String subject,
       previousAvroSchemas.add(previousAvroSchema.schemaObj);
     }
 
-    AvroCompatibilityLevel compatibility = getCompatibilityLevel(subject);
-    if (compatibility == null) {
-      compatibility = getCompatibilityLevel(null);
-    }
+    AvroCompatibilityLevel compatibility = getCompatibilityLevelInScope(subject);
     return compatibility.compatibilityChecker
         .isCompatible(AvroUtils.parseSchema(newSchemaObj).schemaObj, previousAvroSchemas);
   }
 
+  private AvroCompatibilityLevel getCompatibilityLevelInScope(String subject)
+      throws SchemaRegistryStoreException {
+    AvroCompatibilityLevel compatibilityLevel = getCompatibilityLevel(subject);
+    if (compatibilityLevel != null) {
+      return compatibilityLevel;
+    }
+    ConfigValue configValue = lookupCache.configInScope(subject);
+    if (configValue != null) {
+      return configValue.getCompatibilityLevel();
+    }
+    return getCompatibilityLevel(null);
+  }
+
   private void deleteSubjectCompatibility(String subject) throws StoreException {
     ConfigKey configKey = new ConfigKey(subject);
     this.kafkaStore.delete(configKey);
   }
 
+  public Mode getMode(String subject) throws SchemaRegistryStoreException {
+    try {
+      ModeKey modeKey = new ModeKey(subject);
+      ModeValue modeValue = (ModeValue) kafkaStore.get(modeKey);
+      if (modeValue == null && subject == null) {
+        // default global mode
+        return Mode.READWRITE;
+      }
+      return modeValue != null ? modeValue.getMode() : null;
+    } catch (StoreException e) {
+      throw new SchemaRegistryStoreException("Failed to read mode from the kafka store", e);
+    }
+  }
+
+  public Mode getModeInScope(String subject) throws SchemaRegistryStoreException {
+    Mode mode = getMode(subject);
+    if (mode != null) {
+      return mode;
+    }
+    ModeValue modeValue = lookupCache.modeInScope(subject);
+    if (modeValue != null) {
+      return modeValue.getMode();
+    }
+    return getMode(null);
+  }
+
+  public void setMode(String subject, Mode mode)
+      throws SchemaRegistryStoreException, OperationNotPermittedException {
+    if (!allowModeChanges) {
+      throw new OperationNotPermittedException("Mode changes are not allowed");
+    }
+    if (mode == Mode.IMPORT && getMode(subject) != Mode.IMPORT) {
+      // Changing to import mode requires that no schemas exist with matching subjects.
+      Set<String> matchingSubjects = listSubjects(subject);
+      if (!matchingSubjects.isEmpty()) {
```

Comment:
Could there be scenarios where the subject exists with no schema versions?

Reply:
For import I only care that no schema versions exist for the matching subjects.